topology/src/main/java/io/camunda/zeebe/topology/TopologyInitializer.java

```diff
@@ -0,0 +1,271 @@
+import io.camunda.zeebe.topology.state.ClusterChangePlan;
+import io.camunda.zeebe.topology.state.ClusterTopology;
+import io.camunda.zeebe.topology.state.MemberState;
+import io.camunda.zeebe.topology.state.PartitionState;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Initialized topology * */
+public interface TopologyInitializer {
+
+  /**
+   * Initializes the cluster topology.
+   *
+   * @return a future when completed with true indicates that the cluster topology is initialized
+   *     successfully. Otherwise, the topology is not initialized.
+   */
+  ActorFuture<Boolean> initialize();
+
+  /**
+   * Chain initializers in oder. If this.initialize did not successfully initialize, then the
+   * topology is initialized using the provided initializer
+   *
+   * @param after the next initializer used to initialize topology if the current one did not
+   *     succeed.
+   * @return a chained TopologyInitializer
+   */
+  default TopologyInitializer orThen(final TopologyInitializer after) {
+    final TopologyInitializer actual = this;
+    return () -> {
+      final ActorFuture<Boolean> chainedInitialize = new CompletableActorFuture<>();
+      actual
+          .initialize()
+          .onComplete(
+              (initialized, error) -> {
+                if (error != null || !initialized) {
+                  after.initialize().onComplete(chainedInitialize);
+                } else {
+                  chainedInitialize.complete(initialized);
+                }
+              });
+      return chainedInitialize;
+    };
+  }
+
+  /** Initialized topology from the locally persisted topology */
+  class FileInitializer implements TopologyInitializer {
+
+    private final PersistedClusterTopology persistedClusterTopology;
+
+    public FileInitializer(final PersistedClusterTopology persistedClusterTopology) {
+      this.persistedClusterTopology = persistedClusterTopology;
+    }
+
+    @Override
+    public ActorFuture<Boolean> initialize() {
+      try {
+        persistedClusterTopology.tryInitialize();
+        return CompletableActorFuture.completed(!persistedClusterTopology.isUninitialized());
+      } catch (final Exception e) {
+        return CompletableActorFuture.completedExceptionally(e);
+      }
+    }
+  }
+
+  /**
+   * Initializes local topology from the topology received from other members via gossip.
+   * Initialization completes successfully, when it receives a valid initialized topology from any
+   * member. The future returned by initialize is never completed until a valid topology is
+   * received.
+   */
+  class GossipInitializer implements TopologyInitializer, Listener {
```

Comment:
:thought_balloon: Mostly just a naming thing but it looks a bit odd that the _gossip_ initializer actually doesn't do anything with the gossiper but instead listens to updates of the locally persisted topology. Should we instead register listeners on  `ClusterTopologyGossiper` directly?

Reply:
Source of truth is `PersistedClusterTopology`. That's why I added the listener there.